### PR TITLE
*: support flashback dropped/truncated table

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
* Support flashback dropped/truncated table.
* Parser PR: https://github.com/pingcap/parser/pull/611
eg:
```sql
test> create table t (a int key,b int);
test> insert into t values (1,1);
test> truncate table t; -- truncate the table;
test> admin show ddl jobs 3; -- look for the start_time of the truncate job. 
+--------+---------+--------------------+----------------+--------------+-----------+----------+-----------+-----------------------------------+-----------------------------------+--------+
| JOB_ID | DB_NAME | TABLE_NAME         | JOB_TYPE       | SCHEMA_STATE | SCHEMA_ID | TABLE_ID | ROW_COUNT | START_TIME                        | END_TIME                          | STATE  |
+--------+---------+--------------------+----------------+--------------+-----------+----------+-----------+-----------------------------------+-----------------------------------+--------+
| 46     | test    | t                  | truncate table | public       | 1         | 43       | 0         | 2019-11-04 11:56:49.497 +0800 CST | 2019-11-04 11:56:49.598 +0800 CST | synced |
| 44     | test    | t                  | create table   | public       | 1         | 43       | 0         | 2019-11-04 11:56:11.647 +0800 CST | 2019-11-04 11:56:11.747 +0800 CST | synced |
| 42     | mysql   | opt_rule_blacklist | create table   | public       | 3         | 41       | 0         | 2019-11-04 11:48:04.137 +0800 CST | 2019-11-04 11:48:04.19 +0800 CST  | synced |
+--------+---------+--------------------+----------------+--------------+-----------+----------+-----------+-----------------------------------+-----------------------------------+--------+
test> select * from t;
+---+---+
| a | b |
+---+---+
test> flashback table t until timestamp '2019-11-04 11:56:49.497 +0800 CST' to t1;  
test> select * from t1;
+---+---+
| a | b |
+---+---+
| 1 | 1 |
+---+---+
test> insert into t1 values (2,2)
test> select * from t1;
+---+---+
| a | b |
+---+---+
| 1 | 1 |
| 2 | 2 |
+---+---+
test> select * from t;
+---+---+
| a | b |
+---+---+
```

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

Related changes

Release note

 -  Support flashback dropped/truncated table.

### CODE REVIEW

**Comment by djshow832** [comment]
Why don't you use `RECOVER TABLE t UNTIL TIMESTAMP ...`? It can be compatible with `RECOVER TALBE t` and `RECOVER TABLE BY ...`.

File: executor/executor_test.go [hunk 1]
```diff
@@ -4663,6 +4663,108 @@ func (s *testRecoverTable) TestRecoverTable(c *C) {
...
+	ts = getDDLJobStartTime(tk, "test_flashback", "t_flashback2")
+	tk.MustExec(fmt.Sprintf("flashback table t_flashback2 until timestamp '%v' to t_flashback3", ts))
+	// check flashback table meta and data record.
+	tk.MustQuery("select * from t_flashback3;").Check(testkit.Rows("1", "2", "3", "4", "5", "6", "7", "8", "9"))
+	// check flashback table autoID.
+	tk.MustExec("insert into t_flashback3 values (10),(11)")
+	tk.MustQuery("select * from t_flashback3;").Check(testkit.Rows("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11"))
+	// check rebase auto id.
+	tk.MustQuery("select a,_tidb_rowid from t_flashback3;").Check(testkit.Rows("1 1", "2 2", "3 3", "4 5001", "5 5002", "6 5003", "7 10001", "8 10002", "9 10003", "10 15001", "11 15002"))
+}
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/executor_test.go]: Great catch.

**Comment by crazycs520** [comment]
@djshow832 `flashback` syntax is the PM demand.🙃

File: executor/executor_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/executor_test.go]: done.

File: executor/ddl.go [hunk 2]
```diff
@@ -328,23 +331,32 @@ func (e *DDLExec) executeRecoverTable(s *ast.RecoverTableStmt) error {
...
+		job, tblInfo, err = e.getRecoverTableByTableName(s.Table, "")
 	}
 	if err != nil {
 		return err
 	}
+	autoID, err := e.getTableAutoIDFromSnapshot(job)
+	if err != nil {
+		return err
+	}
+	// Call DDL RecoverTable
```
Comment (AilinKid) [inline executor/ddl.go]:
```suggestion
	// Call DDL RecoverTable.
```

File: executor/ddl.go [hunk 3]
```diff
@@ -392,24 +409,28 @@ func (e *DDLExec) getRecoverTableByTableName(s *ast.RecoverTableStmt, t *meta.Me
...
+		if job.Type != model.ActionDropTable && job.Type != model.ActionTruncateTable {
 			continue
 		}
 		// Check GC safe point for getting snapshot infoSchema.
 		err = gcutil.ValidateSnapshotWithGCSafePoint(job.StartTS, gcSafePoint)
 		if err != nil {
 			return nil, nil, err
 		}
+		if len(ts) != 0 && ts != model.TSConvert2Time(job.StartTS).String() {
+			continue
```
Comment (AilinKid) [inline executor/ddl.go]: Why `flushback` syntax requires the `ts` exactly the same with the job's startTS?

File: executor/ddl.go [hunk 4]
```diff
@@ -437,11 +458,39 @@ func (e *DDLExec) getRecoverTableByTableName(s *ast.RecoverTableStmt, t *meta.Me
 		}
 	}
 	if tblInfo == nil {
-		return nil, nil, errors.Errorf("Can't found drop table: %v in ddl history jobs", s.Table.Name)
+		return nil, nil, errors.Errorf("Can't find dropped table: %v in ddl history jobs", tableName.Name)
```
Comment (AilinKid) [inline executor/ddl.go]: We could define a new Errxxx like `infoschema.ErrDatabaseNotExists`

File: executor/ddl.go [hunk 5]
```diff
@@ -437,11 +458,39 @@ func (e *DDLExec) getRecoverTableByTableName(s *ast.RecoverTableStmt, t *meta.Me
...
 	}
 	return job, tblInfo, nil
 }
 
+func (e *DDLExec) executeFlashbackTable(s *ast.FlashBackTableStmt) error {
+	ts := s.Timestamp.GetString()
+	if len(ts) == 0 {
+		return errors.Errorf("The timestamp in flashback statement should be consistent with the drop/truncate DDL start time")
```
Comment (AilinKid) [inline executor/ddl.go]: ditto here.

File: executor/ddl.go [hunk 6]
```diff
@@ -437,11 +458,39 @@ func (e *DDLExec) getRecoverTableByTableName(s *ast.RecoverTableStmt, t *meta.Me
...
+	_, ok := is.TableByID(tblInfo.ID)
+	if ok {
+		return infoschema.ErrTableExists.GenWithStackByArgs("tableID:" + strconv.FormatInt(tblInfo.ID, 10))
+	}
+
+	autoID, err := e.getTableAutoIDFromSnapshot(job)
+	if err != nil {
+		return err
+	}
+	// Call DDL RecoverTable
```
Comment (AilinKid) [inline executor/ddl.go]:
```suggestion
	// Call DDL RecoverTable.
```

File: executor/ddl.go (hunk 3 above)
Comment (crazycs520) [reply executor/ddl.go]: It is the PM demand...

Comment (AilinKid) [reply executor/ddl.go]: OK

File: executor/executor_test.go [hunk 7]
```diff
@@ -4672,6 +4672,131 @@ func (s *testRecoverTable) TestRecoverTable(c *C) {
...
+			ddl.EmulatorGCDisable()
+		}
+	}(ddl.IsEmulatorGCEnable())
+
+	// disable emulator GC.
+	// Otherwise emulator GC will delete table record as soon as possible after execute drop table ddl.
+	ddl.EmulatorGCDisable()
+	gcTimeFormat := "20060102-15:04:05 -0700 MST"
+	timeBeforeDrop := time.Now().Add(0 - time.Duration(48*60*60*time.Second)).Format(gcTimeFormat)
+	safePointSQL := `INSERT HIGH_PRIORITY INTO mysql.tidb VALUES ('tikv_gc_safe_point', '%[1]s', '')
```
Comment (AilinKid) [inline executor/executor_test.go]: `savePoint` shouldn't be mixed with `safePoint`, maybe need a cleanup in next pr.

File: executor/executor_test.go [hunk 8]
```diff
@@ -4672,6 +4672,131 @@ func (s *testRecoverTable) TestRecoverTable(c *C) {
...
+	tk.MustQuery("select * from t_p_flashback order by a;").Check(testkit.Rows("1", "2", "3"))
+	// check flashback table autoID.
+	tk.MustExec("insert into t_p_flashback values (4),(5)")
+	tk.MustQuery("select a,_tidb_rowid from t_p_flashback order by a;").Check(testkit.Rows("1 1", "2 2", "3 3", "4 5001", "5 5002"))
+
+	// Test for flashback truncate partition table.
+	tk.MustExec("truncate table t_p_flashback")
+	ts = getDDLJobStartTime(tk, "test_flashback", "t_p_flashback")
+	tk.MustExec(fmt.Sprintf("flashback table t_p_flashback until timestamp '%v' to t_p_flashback1", ts))
+	// check flashback table meta and data record.
```
Comment (AilinKid) [inline executor/executor_test.go]: Last by the way, uniform the capitalization, `Test xxx` / `check xxx`...

File: executor/ddl.go [hunk 9]
```diff
@@ -437,11 +458,39 @@ func (e *DDLExec) getRecoverTableByTableName(s *ast.RecoverTableStmt, t *meta.Me
...
+	}
+	job, tblInfo, err := e.getRecoverTableByTableName(s.Table, ts)
+	if err != nil {
+		return err
+	}
+	if len(s.NewName) != 0 {
+		tblInfo.Name = model.NewCIStr(s.NewName)
+	}
+	// Check the table ID was not exists.
+	is := domain.GetDomain(e.ctx).InfoSchema()
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/ddl.go]: Ye, here `infoschema` not synced is doesn't matter, ddl owner will check again.

File: executor/executor_test.go (hunk 7 above)
Comment (crazycs520) [reply executor/executor_test.go]: Where is the`savePoint`?

File: executor/executor_test.go (hunk 8 above)
Comment (crazycs520) [reply executor/executor_test.go]: Great, done.

**Pull request merged**